### PR TITLE
Support LFS, sparse, filtered, and nested checkouts

### PR DESCRIPTION
**Pull request by lox**
## Why

Workflows such as this fail checkout validation even though they need only local Git and the event repository's existing read authority:

```yaml
- uses: actions/checkout@v7
  with:
    path: sources/application
    sparse-checkout: src
    lfs: true
```

## What

Support LFS, partial-clone filters, cone and non-cone sparse checkout, nested workspace paths, and `clean: false` for a fresh target. Fetch depth now also reaches submodules unchanged instead of collapsing every positive value to `1`.

Repository, LFS, and submodule credentials remain command-scoped. Alternate repositories, credential persistence, paths outside the workspace, and existing-directory reuse still fail explicitly.

### CODE REVIEW

File: internal/runtime/checkout.go [hunk 1]
```diff
@@ -105,34 +105,64 @@ func (r Runner) runCheckout(ctx context.Context, processor *commandProcessor, wo
...
+	}
+	fetchArgs := checkoutFetchArgs(inputs, job.Event.SHA)
+	if err := run(env, credentialed, fetchArgs...); err != nil {
 		return result, err
 	}
 	checkoutTarget := checkoutRevision(inputs, job.Event.SHA)
-	if err := run(env, "checkout", "--detach", checkoutTarget); err != nil {
+	sparse := checkoutSparsePatterns(inputs)
+	if lfs && len(sparse) == 0 {
+		if err := run(env, credentialed, "lfs", "fetch", "origin", checkoutTarget); err != nil {
```
[the remark this answers is not in the document]
Comment (lox) [reply internal/runtime/checkout.go]: Fixed in 9d85e74. Git LFS is now resolved and canonicalized before action hooks, and install, fetch, worktree filters, and submodule filters invoke that pinned executable. Hook execution remains disabled; `lfs install --skip-repo` configures filters without adding push or locking hooks. PATH-poisoning, quoting, missing-LFS cleanup, and submodule pinning tests cover the boundary.

File: internal/runtime/checkout.go [hunk 2]
```diff
@@ -105,34 +105,64 @@ func (r Runner) runCheckout(ctx context.Context, processor *commandProcessor, wo
...
+	if err := configureSparseCheckout(checkoutDirectory, inputs, func(args ...string) error {
+		return run(env, false, args...)
+	}); err != nil {
+		return result, fmt.Errorf("%s sparse checkout: %w", adapter, err)
+	}
+	checkoutBase := base
+	if lfs {
+		checkoutBase = lfsBase
+	}
+	if err := runWithBase(env, checkoutBase, credentialed && lfs && len(sparse) != 0, "checkout", "--detach", checkoutTarget); err != nil {
```
[the remark this answers is not in the document]
Comment (lox) [reply internal/runtime/checkout.go]: Fixed in 9d85e74. Checkout now receives the command-scoped repository helper whenever the effective partial-clone filter is non-empty, including explicit `filter` and sparse checkout’s implicit `blob:none`. The exact command sequence and helper scope are covered by the runtime test.

**Comment by lox** [comment]
@buildsworth-bk review

**Comment by lox** [comment]
CI needs a Buildkite-org reader: aggregate builds [1777](https://buildkite.com/buildkite/buildkite-gha/builds/1777) and [1780](https://buildkite.com/buildkite/buildkite-gha/builds/1780) both fail after about 2m08s, while every GitHub-exposed hosted subcheck passes. The full local `mise run check` and targeted standard/race checkout tests pass, but this executor cannot access the private failed job or logs. Please paste the failed job name and decisive log lines here so I can address it; I will not guess from the aggregate status.

**Comment by lox** [comment]
Authenticated `bk` and diagnosed builds 1777/1780. CI exposed a real issue: where Git LFS is installed, Git prefers the external `git-lfs` command over a same-named alias, so the prior alias did not pin execution. Fixed in 5c161bb by invoking the pre-resolved Git LFS executable directly and carrying the scoped Git security/credential configuration through `GIT_CONFIG_COUNT`/`GIT_CONFIG_KEY_*`/`GIT_CONFIG_VALUE_*`. Also fixed macOS canonical `/private/tmp` expectations. Targeted standard/race runtime suites and full `mise run check` pass.

File: internal/runtime/checkout.go [hunk 3]
```diff
@@ -421,13 +565,38 @@ func repositoryProviderCheckoutCredentialArgs(base []string, agent, host string)
...
+func (r Runner) runRepositoryProviderCheckoutLFS(ctx context.Context, processor *commandProcessor, workspace string, env map[string]string, gitLFS string, commandArgs []string, credentialHost string) error {
+	credentialEnv, err := r.repositoryProviderCheckoutCredentialEnvironment(processor, env, credentialHost)
+	if err != nil {
+		return err
+	}
+	credentialArgs := repositoryProviderCheckoutCredentialArgs(nil, r.RepositoryCredentials.Agent, credentialHost)
+	credentialEnv = checkoutGitConfigEnvironment(credentialEnv, credentialArgs)
+	cmd := exec.Command(gitLFS, commandArgs...)
+	cmd.Dir = workspace
+	cmd.Env = processEnv(credentialEnv)
```
[the remark this answers is not in the document]
Comment (lox) [reply internal/runtime/checkout.go:603]: Fixed in 123b034. LFS-enabled checkout now restricts PATH to the canonical pre-resolved Git directory followed only by immutable system utility directories, and requires that executable to be absolute and named `git`. This applies to direct LFS commands and filter/submodule descendants. The pre-hook poison test now makes the pinned LFS executable launch `git`, proving the swapped workflow PATH entry is never selected.

File: internal/runtime/checkout.go [hunk 4]
```diff
@@ -421,7 +537,7 @@ func repositoryProviderCheckoutCredentialArgs(base []string, agent, host string)
 	)
 }
 
-func (r Runner) runRepositoryProviderCheckoutFetch(ctx context.Context, processor *commandProcessor, workspace string, env map[string]string, git string, base, fetchArgs []string, credentialHost string) error {
+func (r Runner) runRepositoryProviderCheckoutGit(ctx context.Context, processor *commandProcessor, workspace string, env map[string]string, git string, base, commandArgs []string, credentialHost string) error {
```
[the remark this answers is not in the document]
Comment (lox) [reply internal/runtime/checkout.go:582]: Fixed in 123b034. The DeepSec auth-shape guidance now covers repository and LFS fetches, filtered checkout materialization, and submodule updates under the command-scoped helper.

**Comment by lox** [comment]
Acknowledged for review 5000125793: prior credential findings were addressed and re-reviewed.

Acknowledged for review 5000156455: the stale DeepSec security guidance is updated in 123b034.

Acknowledged for review 5000189934: the final Git LFS executable finding is fixed in 123b034; Buildsworth confirms it is resolved. Human approval remains required for this L3 credential-path change.

**Review by wolfeidau — CHANGES_REQUESTED** [review]
It would be good to refactor this function, currently routing, release metadata, and validation are split between a map and a large switch.

**Comment by lox** [comment]
Acknowledged for review 5003540978: Fixed in db42d3a. Checkout input routing, release generation, and value validation now live together in one table-driven rule map; the validation loop only handles normalization, duplicate detection, rule lookup, generation gating, and predicate execution. The existing release-gate and value-contract tests still cover the behavior, and `mise run check` passes.

**Comment by lox** [comment]
Updated for review 5003540978: the rebased replacement for db42d3a is a0a2eda. The table-driven checkout input rule refactor is unchanged, and the branch now sits on latest `origin/main` at b17678d. Targeted standard/race tests and `mise run check` pass on the rebased head.

**Comment by lox** [comment]
Acknowledged for review 5004181905: Buildsworth found no new issues on a0a2eda; human approval remains required for the credential-bearing checkout changes.

File: internal/runtime/checkout.go [hunk 5]
```diff
@@ -315,12 +406,58 @@ func prepareCheckoutDirectory(workspace string, inputs map[string]string) (strin
...
+	return patterns
+}
+
+func configureSparseCheckout(checkoutDirectory string, inputs map[string]string, run func(...string) error) error {
+	patterns := checkoutSparsePatterns(inputs)
+	if len(patterns) == 0 {
+		return nil
+	}
+	if checkoutInputTrue(checkoutInput(inputs, "sparse-checkout-cone-mode")) || checkoutInput(inputs, "sparse-checkout-cone-mode") == "" {
+		return run(append([]string{"sparse-checkout", "set", "--"}, patterns...)...)
```
[the remark this answers is not in the document]
Comment (lox) [reply internal/runtime/checkout.go]: Fixed in 2da97a4: cone-mode patterns now go to `git sparse-checkout set --stdin` instead of argv, so the admitted 1 MiB input bound no longer runs into the process argument-size limit. The focused test asserts both the exact Git arguments and stdin payload.

**Comment by lox** [comment]
Acknowledged for review 5004200510: Fixed the sparse-checkout input-size mismatch in 2da97a4 by sending cone patterns through `git sparse-checkout set --stdin`. Targeted runtime standard/race tests and `mise run check` pass.

**Comment by lox** [comment]
Acknowledged for review 5004243585: Buildsworth confirmed the sparse-checkout stdin fix and found no new issues on 2da97a4. Human approval remains required for the credential-bearing checkout changes.

**Review by wolfeidau — APPROVED** [review]
👍🏻

**Comment by sj26** [comment]
Released in [v0.34.0](https://github.com/buildkite/buildkite-gha/releases/tag/v0.34.0)
